File: src/backends/pushbullet.rs
use super::common::{Backend, BackendError, SendOption};
use async_trait::async_trait;
use serde::{Deserialize, Serialize};

const API_URL: &str = "https://api.pushbullet.com/v2/pushes";

#[derive(Deserialize, Serialize)]
struct Body {
    #[serde(rename = "type")]
    _type: String,
    title: String,
    body: String,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct PushbulletConfig {
    pub pushbullet: PushbulletBackend,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct PushbulletBackend {
    token: String,
}

#[async_trait]
impl Backend for PushbulletBackend {
    async fn send(&self, msg: &str, title: &str, _option: &SendOption) -> Result<(), BackendError> {
        let body = &Body {
            _type: "note".to_string(),
            title: title.to_string(),
            body: msg.to_string(),
        };
        let req = match surf::post(API_URL)
            .set_header("Access-Token", self.token.to_string())
            .set_header("Content-Type", "application/json")
            .body_form(body)
        {
            Ok(req) => req,
            Err(err) => return Err(BackendError::Message(err.to_string())),
        };
        let res = match req.await {
            Ok(res) => res,
            Err(err) => return Err(BackendError::Message(err.to_string())),
        };
        if res.status() != 200 {
            return Err(BackendError::Message(res.status().to_string()));
        }

        Ok(())
    }
}

File: backends/pushbullet.go
package backends

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
)

func init() {
	backends["pushbullet"] = NewPushbullet()
}

const PUSHBULET_API_URL = "https://api.pushbullet.com/v2/pushes"

type PushbulletConfig struct {
	Token string  `mapstructure:"token" validate:"required"`
}

type Pushbullet struct {
}

func NewPushbullet() BackendInterface {
	return &Pushbullet{}
}

func (*Pushbullet) GetConfig() interface{} {
	return PushbulletConfig{}
}

func (*Pushbullet) Send(configIface interface{}, title string, message string, status *bool) error {
	config, ok := configIface.(PushbulletConfig)
	if !ok {
		return fmt.Errorf("invalid config")
	}
	body := map[string]interface{}{
		"type": "note",
		"title": title,
		"body": message,
	}
	jsonBody, err := json.Marshal(body)
	if err != nil {
		return err
	}
	req, err := http.NewRequest("POST", PUSHBULET_API_URL, bytes.NewBuffer(jsonBody))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", config.Token))
	res, err := http.DefaultClient.Do(req)
	if err != nil {
		return err
	}
	if res.StatusCode < 200 || res.StatusCode >= 300 {
		return fmt.Errorf("pushbullet: %s", res.Status)
	}

	return nil
}
